# Project Euler Problem 18

<b>By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.</b>

                    3
                   7 4
                  2 4 6
                 8 5 9 3
                 
<b>That is, 3 + 7 + 4 + 9 = 23.

Find the maximum total from top to bottom of the triangle below:</b>

                                75
                              95  64
                            17  47  82
                          18  35  87  10
                        20  04  82  47  65
                      19  01  23  75  03  34
                    88  02  77  73  07  63  67
                  99  65  04  28  06  16  70  92
                41  41  26  56  83  40  80  70  33
              41  48  72  33  47  32  37  16  94  29
            53  71  44  65  25  43  91  52  97  51  14
          70  11  33  28  77  73  17  78  39  68  17  57
        91  71  52  38  17  14  91  43  58  50  27  29  48
      63  66  04  68  89  53  67  30  73  16  69  87  40  31
    04  62  98  27  23  09  70  98  73  93  38  53  60  04  23

In [1]:
def load_tri_file(filename):
    nums = [[]]
    with open(filename,'r') as grid:
        i = 0
        for line in grid:
            for num in str.split(line):
                nums[i].append(int(num))
            i += 1
            if i >= 15:
                break
            nums.append([])
    return nums

triangle = load_tri_file('problem_18_tri.txt')
print(triangle)

[[75], [95, 64], [17, 47, 82], [18, 35, 87, 10], [20, 4, 82, 47, 65], [19, 1, 23, 75, 3, 34], [88, 2, 77, 73, 7, 63, 67], [99, 65, 4, 28, 6, 16, 70, 92], [41, 41, 26, 56, 83, 40, 80, 70, 33], [41, 48, 72, 33, 47, 32, 37, 16, 94, 29], [53, 71, 44, 65, 25, 43, 91, 52, 97, 51, 14], [70, 11, 33, 28, 77, 73, 17, 78, 39, 68, 17, 57], [91, 71, 52, 38, 17, 14, 91, 43, 58, 50, 27, 29, 48], [63, 66, 4, 68, 89, 53, 67, 30, 73, 16, 69, 87, 40, 31], [4, 62, 98, 27, 23, 9, 70, 98, 73, 93, 38, 53, 60, 4, 23]]


In [2]:
def get_sol(tri):
    temp = tri
    print("original last row    ", tri[len(tri)-1])
    print("original 2nd last row", tri[len(tri)-2])
    for i in range(len(tri)-2,-1,-1):
        for j in range(0,len(tri[i])):
            #print(i,j,tri[i][j])
            #print(tri[i+1][j])
            #print(tri[i+1][j+1])
            temp[i][j] = temp[i][j] + max(temp[i+1][j], temp[i+1][j+1])
        print("new last row", temp[i])
    return temp[0][0]



import time
print("Starting Timer!\n====================")
start_time = time.time()

print("Maximum Sum is:", get_sol(triangle))

print("====================\n--- %s seconds ---" % (time.time() - start_time))

Starting Timer!
original last row     [4, 62, 98, 27, 23, 9, 70, 98, 73, 93, 38, 53, 60, 4, 23]
original 2nd last row [63, 66, 4, 68, 89, 53, 67, 30, 73, 16, 69, 87, 40, 31]
new last row [125, 164, 102, 95, 112, 123, 165, 128, 166, 109, 122, 147, 100, 54]
new last row [255, 235, 154, 150, 140, 179, 256, 209, 224, 172, 174, 176, 148]
new last row [325, 246, 187, 178, 256, 329, 273, 302, 263, 242, 193, 233]
new last row [378, 317, 231, 321, 354, 372, 393, 354, 360, 293, 247]
new last row [419, 365, 393, 387, 419, 425, 430, 376, 454, 322]
new last row [460, 434, 419, 475, 508, 470, 510, 524, 487]
new last row [559, 499, 479, 536, 514, 526, 594, 616]
new last row [647, 501, 613, 609, 533, 657, 683]
new last row [666, 614, 636, 684, 660, 717]
new last row [686, 640, 766, 731, 782]
new last row [704, 801, 853, 792]
new last row [818, 900, 935]
new last row [995, 999]
new last row [1074]
Maximum Sum is: 1074
--- 0.0029914379119873047 seconds ---


*Thus 1074 is the solution*

By using a dynamic programming solution the *new row* indicates the current bottom row of the triangle after absorbing the previous bottom of the triangle. Then, finally we reach the top of the triangle and again absorb the maximum of the two.

Thus, 1074 is the max path, which is expected since numbers are 2 digits and we add up 15 of them.